engine.py

```diff
@@ -197,6 +211,153 @@ def log_error(self, msg):
+        panel_id = re.sub("\W", "_", panel_id)
+        panel_id = panel_id.lower()
+        self.log_debug("Unique panel id for %s %s -> %s" % (bundle, dialog_name, panel_id))
+        return panel_id 
+        
+    def _panel_factory_callback(self, panel_id):
+        """
+        Given a panel id, generate a panel UI.
+        
+        This method is intended to be executed by nuke itself as a callback
+        that runs when panels are being restored, either at startup or 
+        at UI profile switch.
+        
+        :param panel_id: Unique identifier string for the panel
+        :returns: panel widget object 
+        """
+        # note: Nuke silently consumes any exceptions within the callback system
+        # so add a catch all exception handler around this to make sure
+        # we properly log any output 
+        panel_widget = None
+        try:
+            # note! not using the import as this confuses nuke's calback system 
+            import tk_nuke
+        
+            self.log_debug("Panel Factory Callback: Generating UI for panel id '%s'." % panel_id)
+            
+            # the callback is registered at the same time as the panel callback,
+            # so we can assume that our _panels lookup data structure has a matching
+            # object present
+            panel = self._panels[panel_id]
+            # create a new panel widget
+            panel_widget = tk_nuke.NukePanelWidget(panel["title"], panel_id, panel["widget_class"])
+            # and add it to the current pane context (nuke handles this state)
+            panel_widget.addToPane()
+            
+        except Exception, e:
+            self.log_exception("Could not generate panel UI for panel id '%s'" % panel_id)
+        
+        return panel_widget
+        
+        
+    def register_panel(self, title, bundle, widget_class):
+        """
+        Register a panel so that it can persist across sessions. This needs to be called
+        by an app as part of its init phase in order to be detected by the DCC initial restore 
+        logic. Executing this method will merely register the panel so that the DCC is 
+        aware of its existence and can launch it at startup UI restore if needed.
+        
+        In order to show or focus on a panel, use the show_panel() method instead.
+        
+        :param title: Title to display in the panel tab
+        :param bundle: Bundle (e.g. app) that the panel belongs to
+        :param widget_class: Class to instantiate. The constructor of this class
+                             mustn't take any arguments.        
+        """
+        # generate unique identifier
+        panel_id = self._generate_panel_id(title, bundle)
+        
+        # track registered panels
+        self.log_debug("Registering panel '%s'" % panel_id)
+        self._panels[panel_id] = {"title": title, "bundle": bundle, "widget_class": widget_class}
+
+        # tell nuke how to create this panel
```

Comment:
Glad you asked :)

Nuke will issue a warning and you are left with a blank panel. Would be cool if we could do somehing nicer than just a blank panel, but i think our options are pretty limited!